docs(hpa): How HPA works in CDaaS

tsonny opened the pull request:
Resolves Jira: [CDAAS-1530]


[CDAAS-1530]: https://armory.atlassian.net/browse/CDAAS-1530?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
WIP draft

Commit:
quick spell check fixes

Commit:
Fixed heading for declaritive style

Commit:
Removed known issue now that Armory is HPA aware

Commit:
more clean up and minor edits

Commit:
rewrites - rethinks - rewords

Commit:
need a good example config file for hpa

Commit:
Updated to only concept and support info

Commit:
not needed

Commit:
WIP updates based on dicussion w/Aimee

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+---
+title: "How CD-as-a-Service implements Kubernetes HPA"
```
```suggestion
title: "How CD-as-a-Service Implements Kubernetes HPA"
```

tsonny replied:
I changed the title during self review - if you don't like it I am OK with going with this suggestion - please review updates.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+---
+title: "How CD-as-a-Service implements Kubernetes HPA"
+linktitle: "HPA"
```
```suggestion
linktitle: "Kubernetes HPA"
```

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+title: "How CD-as-a-Service implements Kubernetes HPA"
+linktitle: "HPA"
+description: >
+  Armory CD-as-a-Service supports Kubernetes deployments with HPA.
```
```suggestion
  Learn how Armory Continuous Deployment-as-a-Service supports Kubernetes deployments with Horizontal Pod Autoscaling.
```

For SEO, to capture variations of key subjects in title and description, I like to use CD-as-a-Service in one spot and Continuous Deployment-as-a-Service in the other

tsonny replied:
I made some changes that do not allow me to commit this change - I will make it as my own commit because i like it and struggled with how to SD this topic.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+**Within your deployment strategy**
+
+- Check logs and system health
+- Run custom tests
```
Curious why these lines are here? Looks like maybe you copied from the Webhook-Based Approvals page and forgot to remove?

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+
+Run integration tests in a staging environment.
+- Perform metric tests
+- Run security scanners
```
Again wondering why these lines are here.

tsonny replied:
removed

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+  Armory CD-as-a-Service supports Kubernetes deployments with HPA.
+---
+ ## What is HPA?
+ Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments in reaction to changes in observed metrics. When the deployment load increases, the number of deployed pods is increased to meet the demand. The scale behavior is horizontal, rather than vertical, which means that rather than assigning an existing pod more memory or CPU to handle the incoming traffic, new pods that identical instance ...
```
```suggestion
 Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments in reaction to changes in observed metrics. When the deployment load increases, the number of deployed pods is increased to meet the demand. The scale behavior is horizontal, rather than vertical.  HPA scales a resource by adding or removing identical instances of your app instead of Kubernetes assigning an existing pod more resources to handle the incoming traffic.
```

tsonny replied:
Please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+
+ The sync interval is set by the `kube-controller-manager`. The default interval is 30 seconds. Change the default interval using the `horizontal-pod-autoscaler-sync-period` flag.
+
+The default delay is 3 minutes after an upscale event to allow metrics to stabilize. Change the default upscale delay period using the `horizontal-pod-autoscaler-upscale-delay` flag.
```
I'm not sure this "What is HPA" section is even needed since it basically restates (or almost copies verbatim) what's in the Kuberntes docs. What about replacing section with a  prereq section that points to the relevant K8s docs?

tsonny replied:
Please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+Armory Continuous Deployment-as-a-Service supports the following HorizontalPodAutoscaler API versions:
+- [v1](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v1/)
+- [v2beta 1 and 2](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2beta2/)
+-  [v2](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/)
```
Great that you added the links. What do you think about moving Supported Versions section to a "##" near or at the bottom?

tsonny replied:
I did this in my re-write - totally agree

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+The default delay is 3 minutes after an upscale event to allow metrics to stabilize. Change the default upscale delay period using the `horizontal-pod-autoscaler-upscale-delay` flag.
+ 
+## How HPA works in CD-as-a-Service
+When HPA is configured for an Armory Continuous Deployment-as-a-Service (CD-as-a-Service) deployment the deployment is converted to a `ReplicaSet` and the HPA configuration file is automatically re-written to reference the generated `ReplicaSet` resource. Continuous Deployment-as-a-Service freezes scaling behavior during a deployment.
```
```suggestion
When you use CD-as-a-Service to deploy a `Deployment` resource,  CD-as-a-Service converts the `Deployment` to a `ReplicaSet`. Correspondingly, when you provide an `HPA` to scale a `Deployment` resource, that `HPA` is automatically rewritten to reference the generated `ReplicaSet` resource.
```

tsonny replied:
please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+## How HPA works in CD-as-a-Service
+When HPA is configured for an Armory Continuous Deployment-as-a-Service (CD-as-a-Service) deployment the deployment is converted to a `ReplicaSet` and the HPA configuration file is automatically re-written to reference the generated `ReplicaSet` resource. Continuous Deployment-as-a-Service freezes scaling behavior during a deployment.
+
+Consider a deployment to upgrade from the v1 to v2 version of your application. In this scenario the v1 application is running with 10 replicas and is scaled by HPA (min: 5 pods, max: 15 pods).
```
```suggestion
CD-as-a-Service freezes scaling behavior during a deployment. Consider a deployment to upgrade from the `v1` to `v2` version of your app. In this scenario, the `v1` app is running with 10 replicas and is scaled by `HPA` (`min: 5 pods`, `max: 15 pods`).
```

tsonny replied:
Please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+
+Consider a deployment to upgrade from the v1 to v2 version of your application. In this scenario the v1 application is running with 10 replicas and is scaled by HPA (min: 5 pods, max: 15 pods).
+ - At the beginning of a deployment, the HPA is deleted. Your v1 application has 10 pods but no longer scales between 5 and 15 pods.
+ - CD as a Service deploys the v2 instance of your application with 10 pods (20 total pods between v1 and v2).
```
```suggestion
 - CD-as-a-Service deploys the `v2` instance of your application with 10 pods (20 total pods between `v1` and `v2`).
```

tsonny replied:
fixed the product name consistency throughout during re-write

tsonny replied:
will codify v1/v2 in new commit

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+When HPA is configured for an Armory Continuous Deployment-as-a-Service (CD-as-a-Service) deployment the deployment is converted to a `ReplicaSet` and the HPA configuration file is automatically re-written to reference the generated `ReplicaSet` resource. Continuous Deployment-as-a-Service freezes scaling behavior during a deployment.
+
+Consider a deployment to upgrade from the v1 to v2 version of your application. In this scenario the v1 application is running with 10 replicas and is scaled by HPA (min: 5 pods, max: 15 pods).
+ - At the beginning of a deployment, the HPA is deleted. Your v1 application has 10 pods but no longer scales between 5 and 15 pods.
```
```suggestion
 - At the beginning of a deployment, CD-as-a-Service deletes the `HPA`. Your `v1` app has 10 pods but no longer scales between 5 and 15 pods.
```

tsonny replied:
please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+Consider a deployment to upgrade from the v1 to v2 version of your application. In this scenario the v1 application is running with 10 replicas and is scaled by HPA (min: 5 pods, max: 15 pods).
+ - At the beginning of a deployment, the HPA is deleted. Your v1 application has 10 pods but no longer scales between 5 and 15 pods.
+ - CD as a Service deploys the v2 instance of your application with 10 pods (20 total pods between v1 and v2).
+ - At the end of the deployment, a new HPA resource is created to target the v2 application.
```
```suggestion
 - At the end of the deployment, CD-as-a-Service creates a new `HPA` resource to target the `v2` app.
```

tsonny replied:
please see rewrite

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+ - At the beginning of a deployment, the HPA is deleted. Your v1 application has 10 pods but no longer scales between 5 and 15 pods.
+ - CD as a Service deploys the v2 instance of your application with 10 pods (20 total pods between v1 and v2).
+ - At the end of the deployment, a new HPA resource is created to target the v2 application.
+ - If the deployment needs to be rolled back, CDaaS deletes the v2 application and re-creates the original HPA.
```
```suggestion
 - If you need to roll back your deployment, CD-as-a-Service deletes the `v2` app and recreates the original `HPA`.
```

tsonny replied:
fixed in new commit

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+ - If the deployment needs to be rolled back, CDaaS deletes the v2 application and re-creates the original HPA.
+
+### Supported versions
+Armory Continuous Deployment-as-a-Service supports the following HorizontalPodAutoscaler API versions:
```
```suggestion
CD-as-a-Service supports the following `HorizontalPodAutoscaler` API versions:
```

tsonny replied:
please see rewrite - the fix is there but I cannot commit suggestion.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+- Run custom tests
+
+**After deployment**
+Once CPU utilization drops to 0, the HPA automatically scales the number of replicas down to 1.
```
Not sure why this is here?

tsonny replied:
Removed :)

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
````diff
@@ -0,0 +1,73 @@
+```
+kubectl get hpa
+```
+### Roll back an HPA configured deployment
````
```suggestion
### What happens during a rollback
```

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
````diff
@@ -0,0 +1,73 @@
+kubectl get hpa
+```
+### Roll back an HPA configured deployment
+During a rollback, Armory CDaaS relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to re-create the HPA. 
````
```suggestion
During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to recreate the `HPA`. 
```

tsonny replied:
I updated this without seeing your comment - please re-review :)

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+
+> Do not delete this annotation between deployments. 
+
+Manual or programmatic changes to the HPA configuration file between deployments (for example, changes to the HPA's min/max boundaries) may not be captured during a rollback.
```
```suggestion
CD-as-a-service may not capture any manual or programmatic changes you make to the `HPA` configuration file between deployment.
```

tsonny replied:
Again the rewrite I did over the weekend does not allow the commit please re-review.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,73 @@
+
+Manual or programmatic changes to the HPA configuration file between deployments (for example, changes to the HPA's min/max boundaries) may not be captured during a rollback.
+
+Point to an HPA configuration file that references a Deployment resource in your manifests. An HPA configuration file can be assigned to each individual Deployment resource.
```
Unsure why this paragraph is in the section about the known rollback issue.

What about adding a separate `##` section 
```
## Where to declare your HPA configuration
CD-as-a-Service requires your app's `HPA` be declared in your app's manifest. Provide an `HPA` that references a `Deployment` resource in that manifest. Each `Deployment` resource can have its own `HPA`.
```

tsonny replied:
Please see rewrite

aimeeu replied:
I don't see how/where to declare your HPA configuration anywhere in the doc. A k8s user can create an HPA using kubectl or declaratively in a manifest. [CDaaS requires an HPA manifest](https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5#horfq-where-to-declare-your-hpa-configuration).

```
## Where to declare your HPA configuration

CD-as-a-Service requires your app's HPA be [declared in a manifest](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#creating-the-autoscaler-declaratively).  Each `Deployment` resource can have its own `HPA`.
```

aimeeu commented on content/en/cd-as-a-service/release-notes/known-issues.md:
```diff
@@ -3,10 +3,6 @@ title: Known Issues and Limitations
 
 ---
 
-## Strategy to use with a HorizontalPodAutoscaler
```
Good catch - I'd run removal past Parth or Dan.

tsonny replied:
removed this

aimeeu reviewed (changes requested):
First pass review.

Commit:
Updates to trim down to conceptual review of how CDaaS manages HPA and improve style.

Commit:
fix to apply Aimee requested change .

Commit:
codified v1 v2

Commit:
typo fix

Commit:
Updates based on Aimee review

Commit:
fixed based on Aimee review

Commit:
changes based on Aimee review

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+---
+title: "Kubernetes HPA deployments in CD-as-a-Service"
```
```suggestion
title: "Kubernetes HPA Deployments in CD-as-a-Service"
```
Titles use Capital case ([style guide](https://docs.armory.io/contribute/create/guides/docs-style-guide/#use-capital-case-for-page-titles)

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+  Learn how Armory Continuous Deployment-as-a-Service supports Kubernetes deployments with Horizontal Pod Autoscaling.
+---
+ # What is HPA?
+ Kubernetes Horizontal Pod Autoscaling (`HPA`) scales deployments based on workload. When the load increases, new pods are generated on the deployed cluster. With `HPA`, the cluster is scaled horizontally by increasing the number of deployed pods on the cluster, rather than vertically increasing memory or CPU of an existing pod. When the workload decreases the cluster is scaled back down to the c ...
```
A few proofreading nits
punctuation, active vs passive
```suggestion
 Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments based on workload. When the load increases, Kubernetes generates new pods on the deployed cluster. With HPA, Kubernetes scales the cluster horizontally by increasing the number of deployed pods on the cluster, rather than vertically increasing the memory or CPU of an existing pod. When the workload decreases, Kubernetes scales the cluster down to the configured minimum number of replicas as defined in the deployment file.
```

parth-srivastav replied:
> With HPA, Kubernetes scales the cluster horizontally

With HPA, Kubernetes scales the application horizontally

>When the workload decreases, Kubernetes scales the cluster down

When the workload decreases, Kubernetes scales the application down

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+ # What is HPA?
+ Kubernetes Horizontal Pod Autoscaling (`HPA`) scales deployments based on workload. When the load increases, new pods are generated on the deployed cluster. With `HPA`, the cluster is scaled horizontally by increasing the number of deployed pods on the cluster, rather than vertically increasing memory or CPU of an existing pod. When the workload decreases the cluster is scaled back down to the c ...
+
+ - `HorizontalPodAutoscaler` controls the scale of a deployment and its `ReplicaSet`. 
```
This sentence is verbatim from the Kubernetes doc on Horizonal Pod Autoscaling.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+ - The default interval is modified using the `horizontal-pod-autoscaler-sync-period` flag. 
+ - The default delay is 3 minutes after an upscale event to allow metrics to stabilize. Change the default upscale delay period using the `horizontal-pod-autoscaler-upscale-delay` flag.
+ - With `HPA` configured, the deployment object manages the size of underlying replica sets. `HPA` does not manage rolling updates by manipulating the replication controller. 
+  > `HPA` is not recommended for replication controllers. 
```
Back to my original comment about not really needing to repeat the content of the Kubernetes [Horizontal Pod Autoscaling doc](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). 

As a general rule, we don't repeat what's in third party documentation or tell people how to use third party products. How are lines 11-16 relevant to CDaaS?

The subject of this page is how CD-as-a-Service implements Kubernetes Horizontal Pod Autoscaling. I think the expectation is that the reader should be familiar with K8s HPA before reading this page's content.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+
+To learn more about `HPA`, read these Kubernetes Documentation topics: 
+- [Horizontal Pod Autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+- [HorizontalPodAutoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)
```
How about putting this in a **Before you begin** section similar to what's in the [Slab doc](https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5)?

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+- [HorizontalPodAutoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)
+ 
+# HPA in CD-as-a-Service
+Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HPA` configured deployments to replica sets. The `HPA` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.
```
```suggestion
When you use CD-as-a-Service to deploy a `Deployment` resource,  CD-as-a-Service converts the `Deployment` to a `ReplicaSet`. Correspondingly, when you provide an `HorizontalPodAutoscaler` to scale a `Deployment` resource, that `HorizontalPodAutoscaler` is automatically rewritten to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment. 
```
`Depolyment`, `ReplicaSet`, and `HorizontalPodAutoscaler` are Kubernetes `kind` objects and that `kind` is specified in a Kubernetes manifest.

I created an example manifests and deployment file in a  [Slab doc](https://armory.slab.com/posts/hpa-83crfhl7) to illustrate Dan's summary of what CDaaS does with a `Deployment` and corresponding `HorizontalPodAutoscaler`.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,43 @@
+# HPA in CD-as-a-Service
+Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HPA` configured deployments to replica sets. The `HPA` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.
+
+> To check the current status of the `HorizontalPodAutoscaler` run `kubectl get hpa`.
```
This line seems out of place in this location.

aimeeu reviewed (changes requested):
General comment: Horizontal Pod Autoscaling vs `HorizontalPodAutoscaler`.  If HPA refers to Horizontal Pod Autoscaling, it doesn't have inline code formatting. If HPA refers to `HorizontalPodAutoscaler`, then it would have inline code formatting. However, since this doc refers to both, I'd spell out each time so the reader doesn't get confused as to which you are talking about.

Commit:
Removed K8s specfic content

Commit:
updates based on Aimee review

Commit:
wip

parth-srivastav commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+- [HorizontalPodAutoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)
+ 
+# Horizontal Pod Autoscaling in CD-as-a-Service
+Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HorizontalPodAutoscaler` configured deployments to replica sets. The `HorizontalPodAutoscaler` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.
```
ReplicaSet / ReplicaSets  should not be replica sets

aimeeu replied:
I thought we talked about keeping Dan's original wording here??

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+---
+title: "Kubernetes Horizonal Pod Autoscaler deployments in CD-as-a-Service"
```
Capital case in titles.

Either `HorizontalPodAutoscaler` or Horizontal Pod Autoscaling

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+---
+
+# Before you begin
+Read the Kubernetes Documentation topics: 
```
What do you think about softening this a bit? 
Maybe something like "Make sure you are familiar with using Horizontal Pod Autoscaling in Kubernetes deployments" or "You understand Horizontal Pod Autoscaling in Kubernetes"  or "Before you begin using a `HorizontalPodAutoscaler` in your CD-as-a-Service deployment, you should have a working knowledge of Horizonal Pod Autoscaling in Kubernetes"  and embed a link(s) to the K8s docs?

tsonny replied:
I went with the other suggestion you made and added a before you begin section.

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+
+For example, consider a deployment to upgrade the v1 version of a CD-as-a-Service app to a v2 release. In this scenario the v1 application is running with 10 replicas and is scaled with a minimum set to 5 pods and a maximum set to 15 pods.
+
+1. When the deployment begins the v1 `HorizontalPodAutoscaler` is deleted. The v1 app is running 10 pods, but no longer scales between 5 and 15 pods.
```
```suggestion
1. When the deployment begins, the v1 `HorizontalPodAutoscaler` is deleted. The v1 app is running 10 pods, but no longer scales between 5 and 15 pods.
```

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+> To check the current status of the `HorizontalPodAutoscaler` run `kubectl get hpa`.
+
+## What happens during a rollback
+During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to re-create the `HorizontalPodAutoScaler`. 
```
```suggestion
During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to recreate the `HorizontalPodAutoScaler`. 
```

aimeeu commented on content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md:
```diff
@@ -0,0 +1,36 @@
+During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to re-create the `HorizontalPodAutoScaler`. 
+
+> Do not delete this annotation between deployments. CD-as-a-service may not capture any manual or programmatic changes you make to the `HorizontalPodAutoscaler` configuration file between deployment.
+
```
Where is the section on "Where to declare your HPA configuration" ?? I think the user needs to know they must declare the `HorizontalPodAutoscaler` in a manifest file rather than creating it using `kubectl`.  https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5#horfq-where-to-declare-your-hpa-configuration

Commit:
All review comments

aimeeu committed:
style guide updates

aimeeu committed:
remove inline code around K8s resource names

The K8s docs don't use inline code syntax around Resources
Added subheadings to How... section
Changed replicas back to Pods

aimeeu committed:
minor tweak for continuity

aimeeu committed:
remove redundant text

aimeeu committed:
another nit tweak

aimeeu committed:
add that scaling is frozen at deploy and rollback start

aimeeu committed:
rollback clarification
